umbral/umbral.py

```diff
@@ -120,98 +119,142 @@ def __bytes__(self):
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return s * params.g == v + (h * e)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def original_components(self):
+        return self._point_eph_e, self._point_eph_v, self._bn_sig
+
+    def _reconstruct(self):
         id_cfrag_pairs = list(self.cfrags.items())
         id_0, cfrag_0 = id_cfrag_pairs[0]
         if len(id_cfrag_pairs) > 1:
-            ids = self.cfrags.keys() 
+            ids = self.cfrags.keys()
             lambda_0 = lambda_coeff(id_0, ids)
             e = lambda_0 * cfrag_0.point_eph_e1
             v = lambda_0 * cfrag_0.point_eph_v1
-            
+
             for id_i, cfrag in id_cfrag_pairs[1:]:
                 lambda_i = lambda_coeff(id_i, ids)
                 e = e + (lambda_i * cfrag.point_eph_e1)
                 v = v + (lambda_i * cfrag.point_eph_v1)
         else:
             e = cfrag_0.point_eph_e1
             v = cfrag_0.point_eph_v1
-        
-        return ReconstructedCapsule(e_prime=e, v_prime=v, x=cfrag_0.point_eph_ni)
 
-    def __bytes__(self):
-        self.to_bytes()
-
-
-class ReconstructedCapsule(object):
-    def __init__(self, e_prime, v_prime, x):
-        self.point_eph_e_prime = e_prime
-        self.point_eph_v_prime = v_prime
-        self.point_eph_ni = x
+        self._point_eph_e_prime = e
+        self._point_eph_v_prime = v
+        self._point_noninteractive = cfrag_0.point_eph_ni
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    def _get_contents(self, recp_priv_key: UmbralPrivateKey,
+                      sender_pub_key: UmbralPublicKey,
+                      pre):
         """
-        Instantiate ReconstructedCapsule from serialized data.
+        Reconstructs the Capsule contents from the attached CFrags,
+        opens the Capsule and returns what is inside.
+
+        This will often by a symmetric key.
         """
-        e_prime = Point.from_bytes(data[0:33], curve)
-        v_prime = Point.from_bytes(data[33:66], curve)
-        eph_ni = Point.from_bytes(data[66:99], curve)
+        recp_pub_key = recp_priv_key.get_pub_key(pre.params)
+        self._reconstruct()
 
-        return ReconstructedCapsule(e_prime, v_prime, eph_ni)
+        key = pre.decapsulate_reencrypted(
+            recp_pub_key.point_key, recp_priv_key.bn_key,
+            sender_pub_key.point_key, self
 
-    def to_bytes(self):
+        )
+        return key
+
+    def decrypt(self, recp_priv_key: UmbralPrivateKey,
+                sender_pub_key: UmbralPublicKey,
+                ciphertext: bytes, pre):
         """
-        Serialize the ReconstructedCapsule to a bytestring.
+        Opens the capsule and gets what's inside.
+        
+        We hope that's a symmetric key, which we use to decrypt the ciphertext
+        and return the resulting cleartext.
```

Comment:
Since it seems that now the `decrypt` operation is in the Capsule, and this class is used for both original and reconstructed capsules, then this method should work for both cases. Right now it's really strange that if you want to decrypt a re-encrypted ciphertext you go through `Capsule.decrypt`, but to decrypt an original ciphertext you use `pre.decrypt`. 

Having said that, I really don't like that the decrypt method is offered by the capsule, it's really weird for several reasons, but at this point, I don't want to halt development for an issue like that. If @michwill and @tuxxy are OK with this, then it's fine by me.

Reply:
Just to add something to the last paragraph: note the asymmetry between encrypting through the `PRE.encrypt` interface, but decrypting through `Capsule.decrypt`. Of course we can do it like that, but boy it's weird.

Reply:
Yeah, I'm somewhat in agreement here.

But understand: it was the original asymmetry between `PRE.encrypt` and `PRE.decrypt` which led me to do this.  Notice that, if you go all the way through the lifecycle of PRE'd content using the API, and try to use `PRE.decrypt` at the very end, you'll get "Generic Umbral Error," because `PRE.decrypt` expects *Alice's* private key, not Bob's.

In fact, for my part, I find `PRE.encrypt` more suspect than `Capsule.decrypt`.  If I had to choose, I'd opt for moving `encrypt` to `Capsule` and making it a `classmethod`.

If we do decide to move `Capsule.decrypt` back to `PRE`, perhaps it's a good idea to combine it with the current `decrypt` method; making it so that it can take either a capsule with original components or one with reconstructed components.

However, we have the odd situation where we sometimes have both - in fact, so far, you've suggested that we'll *always* have both (ie, #37).

But I don't really understand how that works with the `from_bytes` method - if you pass the bytestring with `reconstruct=True` (just as previously with `ReconstructedCapsule.from_bytes`), you end up with a `Capsule` with only reconstructed components.  What then?

Reply:
> If we do decide to move Capsule.decrypt back to PRE, perhaps it's a good idea to combine it with the current decrypt method; making it so that it can take either a capsule with original components or one with reconstructed components.

I think that's completely fine. But as you mention in your last paragraph, that only works when either you decrypt an original capsule, or a reconstructed capsule with the original components. I think we should only allow these 2 options. The other possibility (i.e., reconstructed capsule without original components) doesn't allow for validity checking, and with the small change I have in mind, decryption wouldn't even work.

Reply:
OK, so does that mean we need to rethink serialization?

Reply:
Good catch, didn't think about that. Yes, I think we should. Previous approach to serialization kind of assumed that we'd have two different objects: a capsule and a reconstructed capsule. With the current approach, which is only one object that it's "enriched" with cFrags until it can "reconstruct" the additional elements (note: I'm starting to think that "reconstruct" it's not a good metaphor here, but I don't want to open that door again!), then obviously serialization/deserialization should handle the original and new components.

Reply:
Yeah, "reconstruct" is starting to wither, fortunately.

@michwill had a discussion yesterday about this: I'm now picturing a `Capsule` with a bunch of little slots, and when you plug enough `CFrags` into them, it kind of glows an orange-ish color and a little pedestal rises up, with a little fingerprint scanner on it for `Bob` to use (ie, transmit his private key).

How about we start to try the word "activated" instead of "reconstructed"?

So, we now want to serialize all six components (three original, three activated) right?  Do you have any opinions about how to do that or shall I just give it a crack?

Reply:
> How about we start to try the word "activated" instead of "reconstructed"?

That could work, although for the moment, and given that we are currently used to "reconstruct", I suggest to stick to it for a while, and change it when we agree on new naming conventions (which it seems you will discuss at the homecoming). I simply don't want us to get distracted with semantic discussions right now. 

> @michwill had a discussion yesterday about this: I'm now picturing a Capsule with a bunch of little slots, and when you plug enough CFrags into them, it kind of glows an orange-ish color and a little pedestal rises up, with a little fingerprint scanner on it for Bob to use (ie, transmit his private key).

That's a cool idea for a NuCypher KMS video :-)